Spigot-Server-Patches/0358-Custom-plugin-advancement-API.patch

```diff
@@ -0,0 +1,199 @@
++
++public class CraftAdvancementManager implements AdvancementManager {
++    private ListMultimap<Plugin, AdvancementData> pluginAdvancements = LinkedListMultimap.create();
++    
++    public static volatile boolean needsReload = false;
++
++    @Override
++    public void clearAdvancements(Plugin plugin) {
++        pluginAdvancements.removeAll(plugin);
++        needsReload = true;
++    }
++
++    @Override
++    public void registerAdvancement(Plugin plugin, AdvancementData advancementData) {
++        pluginAdvancements.put(plugin, advancementData);
++        needsReload = true;
++    }
++
++    public void inject(Map<MinecraftKey, Advancement.SerializedAdvancement> advancements) {
++        pluginAdvancements.values().forEach(data -> {
++            try {
++                MinecraftKey key = CraftNamespacedKey.toMinecraft(data.getKey());
++                Advancement.SerializedAdvancement value = fromPluginAdvancement(data);
++                advancements.put(key, value);
++            } catch (Exception exception) {
++                System.err.println("Failed to load plugin advancement " + data.getKey());
++                exception.printStackTrace();
++            }
++        });
++    }
++
++    private Advancement.SerializedAdvancement fromPluginAdvancement(AdvancementData pluginAdvancement) {
++        MinecraftKey parentKey = null;
++        if (pluginAdvancement.getParent() != null) {
++            parentKey = CraftNamespacedKey.toMinecraft(pluginAdvancement.getParent());
++        }
++
++        AdvancementData.AdvancementDisplay display = pluginAdvancement.getDisplay();
++        AdvancementData.AdvancementRewards rewards = pluginAdvancement.getRewards();
++        AdvancementDisplay mcDisplay = display == null ? null : convertDisplay(display);
++        AdvancementRewards mcRewards = convertRewards(rewards);
++        // Parent is resolved in a later step in NMS, via Advancement.SerializedAdvancement#a(function<MinecraftKey, Advancement>)
++        return new Advancement.SerializedAdvancement(parentKey, mcDisplay, mcRewards, Maps.newHashMap(), new String[0][]);
++    }
++
++    private AdvancementDisplay convertDisplay(AdvancementData.AdvancementDisplay display) {
++        ItemStack mcItem = CraftItemStack.asNMSCopy(display.getIcon());
++        IChatBaseComponent mcTitle = new ChatComponentText(display.getTitle());
++        IChatBaseComponent mcDescription = new ChatComponentText(display.getDescription());
++        NamespacedKey background = display.getBackground();
++        MinecraftKey mcBackground = background == null ? null : CraftNamespacedKey.toMinecraft(background);
++        AdvancementFrameType frameType = convertFrameType(display.getFrameType());
++        return new AdvancementDisplay(mcItem, mcTitle, mcDescription, mcBackground, frameType, display.showToast(), display.announceToChat(), display.isHidden());
++    }
++
++    private AdvancementRewards convertRewards(AdvancementData.AdvancementRewards rewards) {
```

Comment:
This needs formatting better; Avoid one-liners/missing braces

Reply:
that line or the whole method?

Reply:
This whole method